src/lib.rs

```diff
@@ -48,5 +53,73 @@
 #![cfg_attr(feature="clippy", deny(clippy, unicode_not_nfc, wrong_pub_self_convention,
                                    option_unwrap_used))]
 
-#[test]
-fn place_holder() {}
+
+#[cfg(target_os = "macos")]
+extern crate libc;
+
+mod app;
+pub use app::App;
+
+
+
+#[cfg(target_os = "linux")]
+mod linux;
+#[cfg(target_os = "linux")]
+pub use linux::{install, open};
+
+
+
+#[cfg(target_os = "windows")]
+mod windows;
+#[cfg(target_os = "windows")]
+pub use windows::{install, open};
+
+
+
+#[cfg(target_os = "macos")]
+mod macos;
+#[cfg(target_os = "macos")]
+pub use macos::{install, open};
+
+
+/// Foreign Function call Interface to use this library
+pub mod ffi {
+	#![allow(unsafe_code, unused)]
+
+	use std::ffi::CString;
+	use std::os::raw::c_char;
+	use super::{App, open as rust_open, install as rust_install};
+
+	#[no_mangle]
+	/// open the given URI on this system
+	pub unsafe extern "C" fn open(uri: *mut c_char) -> bool {
+		if let Ok(string_uri) = CString::from_raw(uri).into_string() {
+			rust_open(string_uri)
+		} else {
+			false
+		}
+	}
+
+	#[no_mangle]
+	/// install the given App definition for each scheme URI on the system
+	/// schemes are a comma delimited list of schemes
+	pub unsafe extern "C" fn install(
```

Comment:
I checked those, but they use a plenty of macros and helper functions we are not available outside of safe_core and adding all of those here or writing them up here, bloats that library way beyond proportion. In that case I'd rather say, the crate _shouldn't have_ FFI itself and we'd move that part into safe_core/the future safe_sdk ...